lib/galaxy/objectstore/irods.py

```diff
@@ -0,0 +1,710 @@
+    except Exception:
+        # Toss it back up after logging, we can't continue loading at this point.
+        log.exception("Malformed iRODS ObjectStore Configuration XML -- unable to continue")
+        raise
+
+class CloudConfigMixin(object):
+
+    def _config_to_dict(self):
+        return {
+            'auth': {
+                'username': self.username,
+                'password': self.password,
+            },
+            'resource': {
+                'name': self.resource,
+            },
+            'zone': {
+                'name': self.zone,
+            },
+            'connection': {
+                'host': self.host,
+                'port': self.port,
+            },
+            'cache': {
+                'size': self.cache_size,
+                'path': self.staging_path,
+            }
+        }
+
+class IRODSObjectStore(DiskObjectStore, CloudConfigMixin):
+    """
+    Object store that stores objects as data objects in an iRODS collections. A local cache 
+    exists that is used as an intermediate location for files between Galaxy and iRODS.
+    """
+    
+    store_type = 'irods'
+    
+    #def _get_bucket(self, bucket_name):
+    #def _key_exists(self, rel_path):
+
+    def __init__(self, config, config_dict):
+        super(IRODSObjectStore, self).__init__(config, config_dict)
+
+        auth_dict = config_dict['auth']
+        self.username = auth_dict.get('username')
+        self.password = auth_dict.get('password')
+
+        resource_dict = config_dict['resource']
+        self.resource = resource_dict.get('name')
+
+        zone_dict = config_dict['zone']
+        self.zone = zone_dict.get('name')
+
+        connection_dict = config_dict['connection']
+        self.host = connection_dict.get('host')
+        self.port = connection_dict.get('port')
+
+        cache_dict = config_dict['cache']
+        self.cache_size = cache_dict.get('size', -1)
+        self.staging_path = cache_dict.get('path') or self.config.object_store_cache_path
+
+        extra_dirs = dict((e['type'], e['path']) for e in config_dict.get('extra_dirs', []))
+        self.extra_dirs.update(extra_dirs)
+
+        log.debug("Object cache dir: %s", self.staging_path)
+        log.debug("Job work dir: %s", self.extra_dirs['job_work'])
+
+        self._initialize()
+        
+    def __del__(self):
+        self.session.cleanup()
+
+    def _initialize(self):
+        if irods is None:
+            raise Exception(IRODS_IMPORT_MESSAGE)
+
+        self.home =  "/" + self.zone + "/home/" + self.username
+
+        self.session = self._configure_connection(host=self.host, port=self.port, user=self.username, password=self.password, zone=self.zone)
+        # Clean cache only if value is set in galaxy.ini
+        if self.cache_size != -1:
+            # Convert GBs to bytes for comparison
+            self.cache_size = self.cache_size * 1073741824
+            # Helper for interruptable sleep
+            self.sleeper = Sleeper()
```

Comment:
Is it a problem if multiple threads and processes execute this independently ?

Reply:
I am not sure. Per my chat with @jmchilton, I used s3.py as a template for irods.py. So, some of the code (including this cache cleaner) comes directly from there.

Reply:
It's a reasonable start, but I wouldn't assume that s3.py is perfect or even actively used.

Reply:
Right. The caching mechanism seems to work, but I am not sure about cache cleanup. Will see if I can write a unit test for this.

Reply:
We did talk a while back about moving all of the object store caching into a unified module with a global cache.  If you can, along the way here, identify how that refactoring should happen and maybe even do it that'd be a nice addition over duplicated code.